### PR TITLE
 MAVLinkV2MessageRaw injection support for UDP based Mavlink connections (allowing a kind of ~Mavlink L2 routing layer).

### PR DESCRIPTION
**Pull request by gbin**
Added a read/write layer for the UDP connection to enable the injection of non parsed messages (MAVLinkV2MessageRaw) in MavLink connections  instead of only full blown parsed messages (MavMessage).

### CODE REVIEW

File: build/parser.rs [hunk 1]
```diff
@@ -63,7 +63,7 @@ impl MavProfile {
                         // it is a bitmask
                         if dsp == "bitmask" {
                             // find the corresponding enum
-                            for mut enm in self.enums.values_mut() {
```
Comment (gbin) [inline build/parser.rs:66]: This is just a warning left over from upstream.

File: src/connection/mod.rs [hunk 2]
```diff
@@ -6,12 +6,13 @@ use std::io::{self};
 mod tcp;
 
 #[cfg(feature = "udp")]
-mod udp;
+pub mod udp;
```
Comment (massi31415) [inline src/connection/mod.rs:14]: why making it pub?

Comment (massi31415) [reply src/connection/mod.rs:14]: I see that the only use case is in the integration test maybe we can have something like ```#[cfg(feature = "integration_test")]```

File: src/connection/mod.rs [hunk 3]
```diff
@@ -6,12 +6,13 @@ use std::io::{self};
...
 
 #[cfg(feature = "direct-serial")]
 mod direct_serial;
 
 mod file;
+pub mod routing;
```
Comment (massi31415) [inline src/connection/mod.rs:22]:
```suggestion
#[cfg(feature = "routing")]
pub mod routing;
```

File: src/connection/udp.rs [hunk 4]
```diff
@@ -60,13 +60,13 @@ pub fn udpin<T: ToSocketAddrs>(address: T) -> io::Result<UdpConnection> {
...
 }
 
-struct UdpWrite {
-    socket: UdpSocket,
-    dest: Option<SocketAddr>,
-    sequence: u8,
+pub(crate) struct UdpWrite {
+    pub(crate) socket: UdpSocket,
+    pub(crate) dest: Option<SocketAddr>,
+    pub(crate) sequence: u8,
```
Comment (massi31415) [inline src/connection/udp.rs]:
```suggestion
pub(in crate::connection) struct UdpWrite {
    pub(in crate::connection) socket: UdpSocket,
    pub(in crate::connection) dest: Option<SocketAddr>,
    pub(in crate::connection) sequence: u8,
```

Comment (massi31415) [reply src/connection/udp.rs]: It seems that UdpWrite is only used with connection and maybe shouldn't be exposed outside of that?

File: src/connection/udp.rs [hunk 5]
```diff
@@ -60,13 +60,13 @@ pub fn udpin<T: ToSocketAddrs>(address: T) -> io::Result<UdpConnection> {
...
 }
 
-struct PacketBuf {
+pub(crate) struct PacketBuf {
```
Comment (massi31415) [inline src/connection/udp.rs]:
```suggestion
pub(in crate::connection) struct PacketBuf {
```

maybe same here?

File: src/connection/udp.rs [hunk 6]
```diff
@@ -110,16 +110,16 @@ impl Read for PacketBuf {
     }
 }
 
-struct UdpRead {
-    socket: UdpSocket,
-    recv_buf: PacketBuf,
+pub(crate) struct UdpRead {
+    pub(crate) socket: UdpSocket,
+    pub(crate) recv_buf: PacketBuf,
```
Comment (massi31415) [inline src/connection/udp.rs]:
```suggestion
pub(in crate::connection) struct UdpRead {
    pub(in crate::connection) socket: UdpSocket,
    pub(in crate::connection) recv_buf: PacketBuf,
```

??

File: src/connection/udp.rs [hunk 7]
```diff
@@ -110,16 +110,16 @@ impl Read for PacketBuf {
...
 }
 
 pub struct UdpConnection {
-    reader: Mutex<UdpRead>,
-    writer: Mutex<UdpWrite>,
+    pub(crate) reader: Mutex<UdpRead>,
+    pub(crate) writer: Mutex<UdpWrite>,
```
Comment (massi31415) [inline src/connection/udp.rs]:
```suggestion
    pub(in crate::connection) reader: Mutex<UdpRead>,
    pub(in crate::connection) writer: Mutex<UdpWrite>,
```

File: src/connection/udp.rs [hunk 8]
```diff
@@ -110,16 +110,16 @@ impl Read for PacketBuf {
...
     protocol_version: MavlinkVersion,
-    server: bool,
+    pub(crate) server: bool,
```
Comment (massi31415) [inline src/connection/udp.rs]:
```suggestion
    pub(in crate::connection) server: bool,
```

File: src/lib.rs [hunk 9]
```diff
@@ -370,9 +370,13 @@ pub fn read_v1_msg<M: Message, R: Read>(
 
 const MAVLINK_IFLAG_SIGNED: u8 = 0x01;
 
+const HEADER_SIZE_V2: usize = 9;
+const SIGNATURE_SIZE_V2: usize = 13;
+pub const MAX_SIZE_V2: usize = 1 + HEADER_SIZE_V2 + 255 + 2 + SIGNATURE_SIZE_V2;
```
Comment (massi31415) [inline src/lib.rs]: why it need to be pub?

File: src/connection/mod.rs (hunk 2 above)
Comment (gbin) [reply src/connection/mod.rs:14]: pub as it is needed externally if we want to be able to route, I am adding it to be pub only if we have the routing feature flag on

File: src/connection/mod.rs (hunk 3 above)
Comment (gbin) [reply src/connection/mod.rs:22]: added

File: src/connection/udp.rs (hunk 4 above)
Comment (gbin) [reply src/connection/udp.rs]: good point done

File: src/connection/udp.rs (hunk 5 above)
Comment (gbin) [reply src/connection/udp.rs]: done with pub(super)

File: src/connection/udp.rs (hunk 7 above)
Comment (gbin) [reply src/connection/udp.rs]: done

File: src/connection/udp.rs (hunk 8 above)
Comment (gbin) [reply src/connection/udp.rs]: done

File: tests/raw_udp_loopback_test.rs [hunk 10]
```diff
@@ -0,0 +1,72 @@
+mod test_shared;
+
+#[cfg(all(feature = "std", feature = "udp", feature = "common"))]
+mod test_udp_connections {
+    use crate::test_shared::get_heartbeat_msg;
+    use mavlink::common::MavMessage;
+    use mavlink::connection::routing::RawMavV2Connection;
+    use mavlink::connection::udp::{udpin, udpout};
+    use mavlink::{read_versioned_msg, MAVLinkV2MessageRaw, MavConnection, MavHeader, MAX_SIZE_V2};
```
Comment (massi31415) [inline tests/raw_udp_loopback_test.rs:9]: MAX_SIZE_V2 is private yet used here !

**Review by massi31415 — APPROVED** [review]
LGTM

**Pull request merged**